meta-mbl-apps/recipes-support/swupdate/files/arm-handlers.c

```diff
@@ -0,0 +1,19 @@
+#include <stddef.h>
+#include "swupdate.h"
+#include "handler.h"
+#include <rootfs-handler.h>
+
+
+int rootfsv4_handler_wrapper(struct img_type *img, void *data)
+{
+    return rootfsv4_handler(img, data);
+}
+
+__attribute__((constructor))
+void rootfs_handler_init(void)
+{
+    register_handler("ROOTFSv4"
```

Comment:
Can't we use a `#define` (and potentially some macros) that can be passed at compile time to determine what version we are going to use?

I'm just worried about the maintenance effort when we decide to support multiple versions.

Reply:
I started to write a response to this, but was starting to get really long. I think it would be easier to speak in person. @diego-sueiro I'll try to catch you after lunch.

Reply:
@diego-sueiro and I have spoken about this.

I think the short answer is: not without either:
* Adding another place that knows about image format versions.
* Making things even more complicated.

Reply:
I think there's at least couple of options:
* Register all our existing handlers and use the sw-description to determine which version to use
* Have an ifdef here which selects the handlers to register based on the payload version which is set in a config file somewhere

I'm not convinced it's worth adding ifdefs when swupdate will select the handler automatically based on the sw-description.

Reply:
> 
> 
> @diego-sueiro and I have spoken about this.
> 
> I think the short answer is: not without either:
> 
>     * Adding another place that knows about image format versions.
> 
>     * Making things even more complicated.

Missed this when I was typing my previous response. Agreed.

Reply:
Sorry Rob, saying that Diego and I have spoken about this isn't very useful information for you!

> Register all our existing handlers and use the sw-description to determine which version to use

I don't think that would alleviate the issue that @diego-sueiro has identified - unless we do something else as well, we'd still need to mention all the versions of the handlers in this file. Maybe I'm misunderstanding what you mean...

> Have an ifdef here which selects the handlers to register based on the payload version which is set in a config file somewhere

Right, yeah this is sort-of what I was getting at:
* We could add some config to our BitBake environment. That would add another place that knows about image format versions. We could use it to remove knowledge from this file (arm-handlers.c), but both `create-update-payload` and the swupdate-handlers code would still need to know about all the versions of handlers. We'd still have three places that need updating when we add new handlers.

  We could potentially use config in the BitBake environment to cause build time errors if not all three of `swupdate-handlers`, `swupdate` and `create-update-payload` agree on what the latest versions are, but we've got reasonable automated tests that should catch these issues already anyway. And in some cases we'd get a link error for swupdate if we updated swupdate-handlers and forgot to update here.

* We could try to extract the version information that arm-handlers.c needs from swupdate-handlers or `create-update-payload`, but I think that would get complicated.

Reply:
I'm just worried about maintenance and potential backward incompatibilities.

For the maintenance, when we decide to up-rev any payload version, what source code should we modify?
If this is well documented or even generate build-time errors, I don't see any issues.

Reply:
I'll add some comments where there's non-obvious coupling between code.

Reply:
I've added some comments. Some are in https://github.com/ARMmbed/meta-mbl/pull/811/commits/e328c7cde35a9fdd84e61ef81e7db49508a3490e, most are in the swupdate-handlers code in https://github.com/ARMmbed/mbl-core/pull/261/commits/ac4df7db4161bfe6e1708c047aab4ab8b7b04aee and https://github.com/ARMmbed/mbl-core/pull/261/commits/9305879efe61138ae4d536dd29fb599f4b7e1ff1.